File: samples/javascript_nodejs/13.basic-bot/dialogs/bookingDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { TimexProperty } = require('@microsoft/recognizers-text-data-types-timex-expression');
const { ConfirmPrompt, TextPrompt, WaterfallDialog } = require('botbuilder-dialogs');
const { CancelAndHelpDialog } = require('./cancelAndHelpDialog');
const { DateResolverDialog } = require('./dateResolverDialog');

const CONFIRM_PROMPT = 'confirmPrompt';
const DATE_RESOLVER_DIALOG = 'dateResolverDialog';
const TEXT_PROMPT = 'textPrompt';
const WATERFALL_DIALOG = 'waterfallDialog';

class BookingDialog extends CancelAndHelpDialog {
    constructor(id) {
        super(id || 'bookingDialog');

        this.addDialog(new TextPrompt(TEXT_PROMPT))
            .addDialog(new ConfirmPrompt(CONFIRM_PROMPT))
            .addDialog(new DateResolverDialog(DATE_RESOLVER_DIALOG))
            .addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
                this.destinationStep.bind(this),
                this.originStep.bind(this),
                this.travelDateStep.bind(this),
                this.confirmStep.bind(this),
                this.finalStep.bind(this)
            ]));

        this.initialDialogId = WATERFALL_DIALOG;
    }

    async destinationStep(stepContext) {
        const bookingDetails = stepContext.options;

        if (!bookingDetails.destination) {
            return await stepContext.prompt(TEXT_PROMPT, { prompt: 'Where would you like to travel to?' });
        } else {
            return await stepContext.next(bookingDetails.destination);
        }
    }
    
    async originStep(stepContext) {
        const bookingDetails = stepContext.options;

        bookingDetails.destination = stepContext.result;
        if (!bookingDetails.origin) {
            return await stepContext.prompt(TEXT_PROMPT, { prompt: 'Where are you traveling from?' });
        } else {
            return await stepContext.next(bookingDetails.origin);
        }
    }

    async travelDateStep(stepContext) {
        const bookingDetails = stepContext.options;

        bookingDetails.origin = stepContext.result;
        if (!bookingDetails.travelDate || this.isAmbiguous(bookingDetails.travelDate)) {
            return await stepContext.beginDialog(DATE_RESOLVER_DIALOG, bookingDetails.travelDate);
        } else {
            return await stepContext.next(bookingDetails.travelDate);
        }
    }

    async confirmStep(stepContext) {
        const bookingDetails = stepContext.options;

        bookingDetails.travelDate = stepContext.result;
        const msg = `Please confirm, I have you traveling to: ${bookingDetails.destination} from: ${bookingDetails.origin} on: ${bookingDetails.travelDate}`;

        return await stepContext.prompt(CONFIRM_PROMPT, { prompt: msg });
    }

    async finalStep(stepContext) {
        if (stepContext.result == true) {
            const bookingDetails = stepContext.options;

            return await stepContext.endDialog(bookingDetails);
        } else {
            return await stepContext.endDialog();
        }
    }

    isAmbiguous(timex) {
        const timexPropery = new TimexProperty(timex);
        return !timexPropery.types.has('definite');
    }
}

module.exports.BookingDialog = BookingDialog;

File: generators/java/generators/app/templates/core/src/main/java/BookingDialog.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package <%= packageName %>;

import com.microsoft.bot.builder.MessageFactory;
import com.microsoft.bot.dialogs.DialogTurnResult;
import com.microsoft.bot.dialogs.WaterfallDialog;
import com.microsoft.bot.dialogs.WaterfallStep;
import com.microsoft.bot.dialogs.WaterfallStepContext;
import com.microsoft.bot.dialogs.prompts.ConfirmPrompt;
import com.microsoft.bot.dialogs.prompts.PromptOptions;
import com.microsoft.bot.dialogs.prompts.TextPrompt;
import com.microsoft.bot.schema.Activity;
import com.microsoft.bot.schema.InputHints;
import com.microsoft.recognizers.datatypes.timex.expression.Constants;
import com.microsoft.recognizers.datatypes.timex.expression.TimexProperty;

import java.util.Arrays;
import java.util.concurrent.CompletableFuture;

/**
 * The class containing the booking dialogs.
 */
public class BookingDialog extends CancelAndHelpDialog {

    private final String destinationStepMsgText = "Where would you like to travel to?";
    private final String originStepMsgText = "Where are you traveling from?";

    /**
     * The constructor of the Booking Dialog class.
     */
    public BookingDialog() {
        super("BookingDialog");

        addDialog(new TextPrompt("TextPrompt"));
        addDialog(new ConfirmPrompt("ConfirmPrompt"));
        addDialog(new DateResolverDialog(null));
        WaterfallStep[] waterfallSteps = {
            this::destinationStep,
            this::originStep,
            this::travelDateStep,
            this::confirmStep,
            this::finalStep
        };
        addDialog(new WaterfallDialog("WaterfallDialog", Arrays.asList(waterfallSteps)));

        // The initial child Dialog to run.
        setInitialDialogId("WaterfallDialog");
    }


    private CompletableFuture<DialogTurnResult> destinationStep(WaterfallStepContext stepContext) {
        BookingDetails bookingDetails = (BookingDetails) stepContext.getOptions();

        if (bookingDetails.getDestination().isEmpty()) {
            Activity promptMessage =
                MessageFactory.text(destinationStepMsgText, destinationStepMsgText,
                    InputHints.EXPECTING_INPUT
                );
            PromptOptions promptOptions = new PromptOptions();
            promptOptions.setPrompt(promptMessage);
            return stepContext.prompt("TextPrompt", promptOptions);
        }

        return stepContext.next(bookingDetails.getDestination());
    }


    private CompletableFuture<DialogTurnResult> originStep(WaterfallStepContext stepContext) {
        BookingDetails bookingDetails = (BookingDetails) stepContext.getOptions();

        bookingDetails.setDestination(stepContext.getResult().toString());

        if (bookingDetails.getOrigin().isEmpty()) {
            Activity promptMessage =
                MessageFactory
                    .text(originStepMsgText, originStepMsgText, InputHints.EXPECTING_INPUT);
            PromptOptions promptOptions = new PromptOptions();
            promptOptions.setPrompt(promptMessage);
            return stepContext.prompt("TextPrompt", promptOptions);
        }

        return stepContext.next(bookingDetails.getOrigin());
    }


    private CompletableFuture<DialogTurnResult> travelDateStep(WaterfallStepContext stepContext) {
        BookingDetails bookingDetails = (BookingDetails) stepContext.getOptions();

        bookingDetails.setOrigin(stepContext.getResult().toString());

        if (bookingDetails.getTravelDate() == null || isAmbiguous(bookingDetails.getTravelDate())) {
            return stepContext.beginDialog("DateResolverDialog", bookingDetails.getTravelDate());
        }

        return stepContext.next(bookingDetails.getTravelDate());
    }


    private CompletableFuture<DialogTurnResult> confirmStep(WaterfallStepContext stepContext) {
        BookingDetails bookingDetails = (BookingDetails) stepContext.getOptions();

        bookingDetails.setTravelDate(stepContext.getResult().toString());

        String messageText =
            String.format(
                "Please confirm, I have you traveling to: %s from: %s on: %s. Is this correct?",
                bookingDetails.getDestination(), bookingDetails.getOrigin(),
                bookingDetails.getTravelDate()
            );
        Activity promptMessage = MessageFactory
            .text(messageText, messageText, InputHints.EXPECTING_INPUT);

        PromptOptions promptOptions = new PromptOptions();
        promptOptions.setPrompt(promptMessage);

        return stepContext.prompt("ConfirmPrompt", promptOptions);
    }


    private CompletableFuture<DialogTurnResult> finalStep(WaterfallStepContext stepContext) {
        if ((Boolean) stepContext.getResult()) {
            BookingDetails bookingDetails = (BookingDetails) stepContext.getOptions();
            return stepContext.endDialog(bookingDetails);
        }

        return stepContext.endDialog(null);
    }

    private static boolean isAmbiguous(String timex) {
        TimexProperty timexProperty = new TimexProperty(timex);
        return !timexProperty.getTypes().contains(Constants.TimexTypes.DEFINITE);
    }
}
